### PR TITLE
LibJS+LibUnicode: Begin implementing Intl.DisplayNames

### PR DESCRIPTION
**Pull request by trflynn89**
tl;dr:
```javascript
> var en = new Intl.DisplayNames("en-US", {"type": "region" });
[Intl.DisplayNames]
  locale: "en"
  type: "region"
  style: "long"
  fallback: "code"
> en.of("US");
"United States"

> var es = new Intl.DisplayNames("es-419", {"type": "region" });
[Intl.DisplayNames]
  locale: "es-419"
  type: "region"
  style: "long"
  fallback: "code"
> es.of("US");
"Estados Unidos"
```

First, this PR downloads and parses the [Unicode CLDR database](https://github.com/unicode-org/cldr-json) within LibUnicode, and adds a code generator to compile the subset of this (rather large) database that we need for ECMA-402. There's quite a bit more we'll need to parse for full compliance, but this is a start.
  
Next, this adds `Intl.DisplayNames` to LibJS. For `Intl.DisplayNames.prototype.of`, only `type` options of `region` are really supported. LibUnicode currently only parses regional data from the CLDR database. Once it parses the data for other `type` options, it will be trivial for `LibJS` to get that data.

Worth noting to avoid confusion: what ECMA-402 refers to as regions is what Unicode TR-35 (usually) refers to as territories. So for the most part, this PR uses "territory" within LibUnicode and "region" within LibJS.

### CODE REVIEW

File: Userland/Libraries/LibUnicode/CodeGenerators/GenerateUnicodeLocale.cpp [hunk 1]
```diff
@@ -0,0 +1,372 @@
...
+static void generate_unicode_locale_header(Core::File& file, UnicodeLocaleData& locale_data)
+{
+    StringBuilder builder;
+    SourceGenerator generator { builder };
+
+    auto generate_enum = [&](StringView name, Vector<String>& values) {
+        quick_sort(values);
+
+        generator.set("name", name);
+        generator.set("underlying", String::formatted("{}UnderlyingType", name));
```
Comment (alimpfard) [inline Userland/Libraries/LibUnicode/CodeGenerators/GenerateUnicodeLocale.cpp:145]: You don't need this, the compiler can tell you that: `UnderlyingType<T>`

File: Userland/Libraries/LibUnicode/Locale.cpp [hunk 2]
```diff
@@ -0,0 +1,205 @@
...
+{
+#if ENABLE_UNICODE_DATA
+    static auto const& available_locales = Detail::available_locales();
+
+    auto it = available_locales.find(locale);
+    if (it == available_locales.end())
+        return {};
+
+    if (auto territory = Detail::territory_from_string(code); territory.has_value())
+        return it->value.territories[static_cast<TerritoryUnderlyingType>(*territory)];
```
Comment (alimpfard) [inline Userland/Libraries/LibUnicode/Locale.cpp:197]:
```suggestion
        return it->value.territories[to_underlying(*territory)];
```

File: Userland/Libraries/LibUnicode/Locale.cpp [hunk 3]
```diff
@@ -0,0 +1,205 @@
...
+}
+
+Optional<LanguageID> parse_unicode_language_id(StringView language)
+{
+    // https://unicode.org/reports/tr35/#Unicode_language_identifier
+    // unicode_language_id = "root"
+    //                       | (unicode_language_subtag (sep unicode_script_subtag)?
+    //                           | unicode_script_subtag)
+    //                         (sep unicode_region_subtag)?
+    //                         (sep unicode_variant_subtag)* ;
```
Comment (alimpfard) [inline Userland/Libraries/LibUnicode/Locale.cpp:76]: This leaves a lot to be desired in alignment :P

Comment (alimpfard) [reply Userland/Libraries/LibUnicode/Locale.cpp:76]: (even if it's a verbatim copy of the spec...)

File: Userland/Libraries/LibUnicode/Locale.cpp [hunk 4]
```diff
@@ -0,0 +1,205 @@
...
+
+    if (language == "root"sv) {
+        language_id.is_root = true;
+        return language_id;
+    }
+
+    auto segments = language.split_view_if([](char ch) {
+        // sep = [-_] ;
+        return (ch == '-') || (ch == '_');
+    },
```
Comment (alimpfard) [inline Userland/Libraries/LibUnicode/Locale.cpp:87]: I think we have `is_any_of("-_"sv)` that returns this lambda.

File: Userland/Libraries/LibUnicode/Locale.cpp [hunk 5]
```diff
@@ -0,0 +1,205 @@
...
+    if (it == available_locales.end())
+        return {};
+
+    if (auto territory = Detail::territory_from_string(code); territory.has_value())
+        return it->value.territories[static_cast<TerritoryUnderlyingType>(*territory)];
+
+    return {};
+#else
+    return {};
+#endif
```
Comment (alimpfard) [inline Userland/Libraries/LibUnicode/Locale.cpp:202]: (weak suggestion, just a style thing)
```suggestion

#endif 
    return {};
```

**Review by alimpfard — COMMENTED** [review]
I'll leave the LibJS bits to someone that likes to review them, but here's a few notes about the other parts.

File: Userland/Libraries/LibUnicode/Locale.cpp (hunk 3 above)
Comment (trflynn89) [reply Userland/Libraries/LibUnicode/Locale.cpp]: > (even if it's a verbatim copy of the spec...)

exactly that :laughing:

File: Userland/Libraries/LibUnicode/Locale.cpp (hunk 4 above)
Comment (trflynn89) [reply Userland/Libraries/LibUnicode/Locale.cpp:88]: Do we? We have `StringView::is_one_of` but that wouldn't work as the unary predicate here

Comment (alimpfard) [reply Userland/Libraries/LibUnicode/Locale.cpp:88]: Here it is: 

https://github.com/SerenityOS/serenity/blob/cf271183b4befeba7c96bd6ca25246b899e836bb/AK/GenericLexer.h#L218

Comment (trflynn89) [reply Userland/Libraries/LibUnicode/Locale.cpp]: Ah, nice!

**Comment by linusg** [comment]
Ref #9633
